### PR TITLE
fix: template not validated properly with json schema

### LINKED ISSUE #176: JsonSchema: Template is not being validated properly
The generated schema is too permissive and allows anything inside the `Resources` property. This is because the CDK schema inherits from a static CloudFormation schema, in which `Resources` accepts `{}`, which means "anything":

https://github.com/cdklabs/decdk/blob/c00f4d6b2e23705b45427182063daf5a6fe644d3/cloudformation.schema.json#L121-L131

We should change this pattern to accept any CloudFormation resource (maybe anything whose `Type` starts with `AWS::`?). In any case, we need to make sure that, in case of invalid templates, the error messages make sense to the user.

### PR DESCRIPTION
**Pull request by mrgrain**
Fixes #176 

Originally I just wanted to remove the jsonschema from the runtime. This is not required anymore, since the type resolution will provide the same feature.

But I didn't just want to remove it, so I added a generic test case to validate all our examples.

This lead me to fixing two bugs with the json schema that prevented proper validation.
The first bug caused everything to just pass validation. The second bug caused Constructs without props to not be recognized as valid.

### CODE REVIEW

File: cloudformation.schema.json [hunk 1]
```diff
@@ -2,6 +2,19 @@
     "$schema": "http://json-schema.org/draft-04/schema#",
     "additionalProperties": false,
     "definitions": {
+        "CfnResource": {
+            "additionalProperties": true,
+            "properties": {
+                "Type": {
+                    "type": "string",
+                    "pattern": "^::$"
```
Comment (mrgrain) [inline cloudformation.schema.json:10]: This is the same condition we currently use in the parser/evaluator.

File: cloudformation.schema.json [hunk 2]
```diff
@@ -2,6 +2,19 @@
     "$schema": "http://json-schema.org/draft-04/schema#",
     "additionalProperties": false,
     "definitions": {
+        "CfnResource": {
+            "additionalProperties": true,
+            "properties": {
+                "Type": {
+                    "type": "string",
+                    "pattern": ".*::.*"
```
Comment (mrgrain) [inline cloudformation.schema.json:10]: This is the same condition we currently use in the parser/evaluator.

File: examples/pure-cfn.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
     "Hello": {
       "Type": "aws-cdk-lib.aws_sqs.Queue",
       "Properties": {
-        "encryption": "Unencrypted"
+        "encryption": "UNENCRYPTED"
```
Comment (mrgrain) [inline examples/pure-cfn.json:7]: This was wrong in the current file

File: src/cdk-schema.ts [hunk 4]
```diff
@@ -191,10 +197,17 @@ function unpackConstruct(
   if (!klass.initializer || klass.abstract) {
     return undefined;
   }
-  if (klass.initializer.parameters.length < 3) {
+  if (klass.initializer.parameters.length < 2) {
     return undefined;
   }
 
+  if (klass.initializer.parameters.length == 2) {
```
Comment (mrgrain) [inline src/cdk-schema.ts:204]: Allow Constructs without props

**Review by lejiati — APPROVED** [review]

File: test/schema/examples.test.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { testExamples, Testing } from '../util';
+
+let typeSystem: reflect.TypeSystem;
+let schema: Schema;
+beforeAll(async () => {
+  typeSystem = await Testing.typeSystem;
+  schema = renderFullSchema(typeSystem);
+});
+
+testExamples(async (example) => {
```
Comment (rix0rrr) [inline test/schema/examples.test.ts:15]: We currently have:

* A set of templates that we test for syntatically valid syntax
* A completely separate set of tests that check that some template syntax semantically parses to something else

It concerns me that we have no guarantee that we test all possible forms completely. Who is to say that the set of syntax checks overlaps completely with the set of semantic checks? 

Can we do a JSON check in every parsed template in the test of unit tests, or something similar, to make sure that we 100% know that our coverage of schema validation overlaps with the coverage of features?

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Pull request merged**